packages/sources/coinpaprika-test/src/crypto-utils.ts

```diff
@@ -0,0 +1,141 @@
+import { AdapterConfig } from '@chainlink/external-adapter-framework/config'
+import { customSettings, getApiEndpoint, getApiHeaders } from './config'
+import { SingleNumberResultResponse } from '@chainlink/external-adapter-framework/util'
+
+export const inputParameters = {
+  base: {
+    aliases: ['from', 'coin'],
+    description: 'The symbol of symbols of the currency to query',
+    required: true,
+    type: 'string',
+  },
+  quote: {
+    aliases: ['to', 'market'],
+    description: 'The symbol of the currency to convert to',
+    required: true,
+    type: 'string',
+  },
+  coinid: {
+    description: 'The coin ID (optional to use in place of `base`)',
+    required: false,
+    type: 'string',
+  },
+} as const
+
+export interface CryptoRequestBody {
+  quotes: string
+}
+
+interface CoinInfo {
+  price: number
+  volume_24h: number
+  volume_24h_change_24h: number
+  market_cap: number
+  market_cap_change_24h: number
+  percent_change_15m: number
+  percent_change_30m: number
+  percent_change_1h: number
+  percent_change_6h: number
+  percent_change_12h: number
+  percent_change_24h: number
+  percent_change_7d: number
+  percent_change_30d: number
+  percent_change_1y: number
+  ath_price: number
+  percent_from_price_ath: number
+}
+
+export interface CryptoResponseSchema {
+  id: string
+  name: string
+  symbol: string
+  rank: number
+  circulating_supply: number
+  total_supply: number
+  max_supply: number
+  beta_value: number
+  first_data_at: string
+  last_updated: string
+  quotes: {
+    [key: string]: CoinInfo
+  }
+  cost?: number
+}
+
+export interface CryptoRequestParams {
+  coinid?: string
+  base: string
+  quote: string
+}
+
+export type EndpointTypes = {
+  Request: {
+    Params: CryptoRequestParams
+  }
+  Response: SingleNumberResultResponse
+  CustomSettings: typeof customSettings
+  Provider: {
+    RequestBody: CryptoRequestBody
+    ResponseBody: CryptoResponseSchema[]
+  }
+}
+
+const chunkArray = (params: CryptoRequestParams[], size = 3): CryptoRequestParams[][] =>
+  params.length > size ? [params.slice(0, size), ...chunkArray(params.slice(size), size)] : [params]
+
+export const buildBatchedRequestBody = (
+  params: CryptoRequestParams[],
+  config: AdapterConfig<typeof customSettings>,
+) => {
+  // Coinpaprika supports up to 3 tickers in a single request, so we need to slice it to have 3 element chunks
+  const chunkedMatrix = chunkArray(params)
+
+  return chunkedMatrix.map((chunkedParams) => ({
+    params: chunkedParams,
+    request: {
+      baseURL: getApiEndpoint(config),
+      url: 'v1/tickers',
+      method: 'GET',
+      headers: getApiHeaders(config),
+      params: {
+        quotes: [...new Set(chunkedParams.map((p) => p.quote.toUpperCase()))].join(','),
+      },
+    },
+  }))
+}
```

Comment:
This part needs to be re-worked a bit.

Currently, this will split up the params set into chunks of 3 _price pair_ requests. However, we should split it into chunks of 3 _quotes_.

Currently, the following input leads to the following output:
```
[ ETH/USD, BTC/USD, LINK/USD, BUSD/USD ] -> [ [ ETH/USD, BTC/USD, LINK/USD ], [ BUSD/USD ] ]
```

Instead, these should all be sent in the same request (`/USD`) - and it should be able to support 2 more unique quotes until it splits requests up into a new chunk.

Reply:
Yes, you are absolutely right. Somehow i lost the context and thought about limit of 3 requests. I will change this.